### PR TITLE
ssh/tailssh: fix mosh over Tailscale SSH

### PR DESCRIPTION
**Pull request by maisem**
Updates #4919

Co-authored-by: James Tucker <raggi@tailscale.com>
Co-authored-by: Brad Fitzpatrick <bradfitz@tailscale.com>
Signed-off-by: Maisem Ali <maisem@tailscale.com>

### CODE REVIEW

File: ssh/tailssh/tailssh.go [hunk 1]
```diff
@@ -1004,14 +1005,20 @@ func (ss *sshSession) run() {
...
 	go func() {
+		defer ss.stdin.Close()
 		_, err := io.Copy(rec.writer("i", ss.stdin), ss)
 		if err != nil {
 			// TODO: don't log in the success case.
 			logf("stdin copy: %v", err)
 		}
-		ss.stdin.Close()
+		if f, ok := ss.stdin.(*os.File); ok {
+			unix.IoctlSetInt(int(f.Fd()), unix.TIOCSBRK, 0)
```
Comment (raggi) [inline ssh/tailssh/tailssh.go:1015]: If we later import https://github.com/pkg/term we can use https://github.com/pkg/term/blob/master/termios/termios_linux.go#L35 which also has fall-back implementations for bsd's, solaris, etc.

Comment (maisem) [reply ssh/tailssh/tailssh.go:1015]: Yeah that's a good find. I created a track bug to investigate that.

**Review by bradfitz — COMMENTED** [review]
Could use some comments :)

File: ssh/tailssh/incubator.go [hunk 2]
```diff
@@ -310,15 +310,22 @@ func (ss *sshSession) launchProcess() error {
...
-	go resizeWindow(pty, winCh)
-	ss.stdout = pty // no stderr for a pty
-	ss.stdin = pty
+	// We need to handle closing the the stdin and stdout separately from each other;
+	// to do this we dup the pty FD and assign one to stdin and one to stdout.
+	ptyDup, err := syscall.Dup(int(pty.Fd()))
+	if err != nil {
+		return err
+	}
+	go resizeWindow(ptyDup, winCh)
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:319]: is ptyDup arbitrary here? Maybe a `/* arbitrary side */` comment?

File: ssh/tailssh/incubator.go [hunk 3]
```diff
@@ -310,15 +310,22 @@ func (ss *sshSession) launchProcess() error {
 	if err != nil {
 		return err
 	}
-	go resizeWindow(pty, winCh)
-	ss.stdout = pty // no stderr for a pty
-	ss.stdin = pty
+	// We need to handle closing the the stdin and stdout separately from each other;
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:313]: typo the the. I'd say:

"We need to be able to close stdin and stdout separately later so make a dup."

File: ssh/tailssh/incubator.go [hunk 4]
```diff
@@ -310,15 +310,22 @@ func (ss *sshSession) launchProcess() error {
...
+	// to do this we dup the pty FD and assign one to stdin and one to stdout.
+	ptyDup, err := syscall.Dup(int(pty.Fd()))
+	if err != nil {
+		return err
+	}
+	go resizeWindow(ptyDup, winCh)
+	ss.stdout = os.NewFile(uintptr(ptyDup), pty.Name())
+	ss.stdin = pty // no stderr for a pty
```
Comment (bradfitz) [inline ssh/tailssh/incubator.go:321]: why's comment on this line now? made more sense by stdout before.

Maybe `ss.stderr = nil // none for a pty` ?

**Review by bradfitz — APPROVED** [review]

File: ssh/tailssh/tailssh.go [hunk 5]
```diff
@@ -1004,20 +1006,22 @@ func (ss *sshSession) run() {
...
 	go func() {
-		_, err := io.Copy(rec.writer("i", ss.stdin), ss)
-		if err != nil {
-			// TODO: don't log in the success case.
+		defer ss.stdin.Close()
+		if _, err := io.Copy(rec.writer("i", ss.stdin), ss); err != nil {
 			logf("stdin copy: %v", err)
+			ss.ctx.CloseWithError(err)
+		} else if ss.ptyReq != nil {
+			ss.stdin.Write([]byte{4}) // EOT
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:1014]:
```
const EOT = 4 // https://en.wikipedia.org/wiki/End-of-Transmission_character
ss.stdin.Write([]byte{EOT})
```

?

File: ssh/tailssh/tailssh.go [hunk 6]
```diff
@@ -1029,13 +1033,15 @@ func (ss *sshSession) run() {
...
 	}
+
 	err = ss.cmd.Wait()
 	// This will either make the SSH Termination goroutine be a no-op,
 	// or itself will be a no-op because the process was killed by the
 	// aforementioned goroutine.
 	ss.exitOnce.Do(func() {})
 
 	if err == nil {
+		ss.logf("Session complete")
```
Comment (bradfitz) [inline ssh/tailssh/tailssh.go:1044]: this isn't redundant? I thought we already logged something when it was done?

**Review by bradfitz — APPROVED** [review]

**Pull request merged**